code/hsec-tools/src/Security/Advisories/Parse.hs

```diff
@@ -150,82 +150,181 @@ parseAdvisory policy attrs raw = do
-  related <-
-    fromMaybe []
-      <$> optional advisory "related" (isArrayOf isString)
-
-  affected <- mandatory table "affected" (isArrayOf parseAffected)
-  references <- mandatory table "references" (isArrayOf parseReference)
-
-  pure $ Advisory
-    { advisoryId = identifier
-    , advisoryModified = modified
-    , advisoryPublished = published
-    , advisoryCWEs = cats
-    , advisoryKeywords = kwds
-    , advisoryAliases = aliases
-    , advisoryRelated = related
-    , advisoryAffected = affected
-    , advisoryReferences = references
-    , advisoryPandoc = doc
-    , advisoryHtml = html
-    , advisorySummary = summary
-    , advisoryDetails = details
-    }
-
-parseAffected :: TOML.Value -> TableParser Affected
-parseAffected v = do
-  tbl <- isTable v
-  package <- mandatory tbl "package" isString
-  cvss <- mandatory tbl "cvss" isString -- TODO validate CVSS format
-  os <- optional tbl "os" $ isArrayOf (isString >=> operatingSystem)
-  arch <- optional tbl "arch" $ isArrayOf (isString >=> architecture)
-  decls <- maybe [] Map.toList <$> optional tbl "declarations" (isTableOf versionRange)
-
-  versions <- mandatory tbl "versions" (isArrayOf parseAffectedVersionRange)
-
-  pure $ Affected
-    { affectedPackage = package
-    , affectedCVSS = cvss
-    , affectedVersions = versions
-    , affectedArchitectures = arch
-    , affectedOS = os
-    , affectedDeclarations = decls
-    }
-
-parseAffectedVersionRange :: TOML.Value -> TableParser AffectedVersionRange
-parseAffectedVersionRange v = do
-  tbl <- isTable v
-  introduced <- mandatory tbl "introduced" isString
-  fixed <- optional tbl "fixed" isString
-  pure $ AffectedVersionRange introduced fixed
+  -> Toml.Table
+  -> Toml.Result Advisory
+parseAdvisoryTable oob policy doc summary details html tab =
+  Toml.runMatcher $
+   do fm <- Toml.fromValue (Toml.Table tab)
+      published <-
+        mergeOobMandatory policy
+          (oobPublished oob)
+          "advisory.published"
+          (amdPublished (frontMatterAdvisory fm))
+      modified <-
+        fromMaybe published <$>
+          mergeOobOptional policy
+            (oobPublished oob)
+            "advisory.modified"
+            (amdModified (frontMatterAdvisory fm))
+      pure Advisory
+        { advisoryId = amdId (frontMatterAdvisory fm)
+        , advisoryPublished = published
+        , advisoryModified = modified
+        , advisoryCWEs = amdCWEs (frontMatterAdvisory fm)
+        , advisoryKeywords = amdKeywords (frontMatterAdvisory fm)
+        , advisoryAliases = amdAliases (frontMatterAdvisory fm)
+        , advisoryRelated = amdRelated (frontMatterAdvisory fm)
+        , advisoryAffected = frontMatterAffected fm
+        , advisoryReferences = frontMatterReferences fm
+        , advisoryPandoc = doc
+        , advisoryHtml = html
+        , advisorySummary = summary
+        , advisoryDetails = details
+        }
+
+data FrontMatter = FrontMatter {
```

Comment:
Could you add a comment that this is an intermediate type to assist with parsing, and that it should not be exported?